src/tpos/tposutils.cpp

```diff
@@ -374,105 +360,250 @@ bool TPoSUtils::IsMerchantPaymentValid(CValidationState &state, const CBlock &bl
-                }
-                else if(txOut.nValue == TPOS_CONTRACT_COLATERAL)
-                {
+                } else if(txOut.nValue == TPOS_CONTRACT_COLATERAL) {
                     colateralFound = true;
                 }
             }
 
-            if(metadataOutPtr && colateralFound)
-            {
+            if (metadataOutPtr && colateralFound) {
                 const auto &metadataOut = *metadataOutPtr;
                 std::vector<std::vector<unsigned char>> vSolutions;
                 txnouttype whichType;
-                if (Solver(metadataOut.scriptPubKey, whichType, vSolutions) && whichType == TX_NULL_DATA)
-                {
+                if (Solver(metadataOut.scriptPubKey, whichType, vSolutions) && whichType == TX_NULL_DATA) {
                     // Here we can have a chance that it is transaction which is a tpos contract, let's check if it has
                     std::stringstream stringStream(metadataOut.scriptPubKey.ToString());
 
-                    std::string tokens[5];
-                    for(auto &token : tokens)
-                    {
-                        stringStream >> token;
+                    std::vector<std::string> tokens;
+                    std::string token;
+                    while(stringStream >> token) {
+                        tokens.emplace_back(token);
                     }
 
-                    CBitcoinAddress tposAddress(ParseAddressFromMetadata(tokens[1]));
-                    CBitcoinAddress merchantAddress(ParseAddressFromMetadata(tokens[2]));
-                    int commission = std::stoi(tokens[3]);
-                    std::vector<unsigned char> vchSignature = ParseHex(tokens[4]);
-                    if(tokens[0] == GetOpName(OP_RETURN) && tposAddress.IsValid() && merchantAddress.IsValid() &&
-                            commission > 0 && commission < 100)
-                    {
-
-                        // if we get to this point, it means that we have found tpos contract that was created for us to act as merchant.
-                        return TPoSContract(tx, merchantAddress, tposAddress, commission, vchSignature);
+                    if(tokens.at(0) == GetOpName(OP_RETURN)) {
+
+                        if (tokens.size() == 2) {
+                            // check for new contract
+                            CDataStream ds(ParseHex(tokens.at(1)), SER_NETWORK, PROTOCOL_VERSION);
+                                    TPoSContract contract;
+                            ds >> contract;
+                            contract.txContract = tx;
+                            if (contract.IsValid()) {
+                                return contract;
+                            }
+                        } else if (tokens.size() > 3 && tokens.size() < 6) {
+
+                            CBitcoinAddress tposAddress(ParseAddressFromMetadata(tokens[1]));
+                            CBitcoinAddress merchantAddress(ParseAddressFromMetadata(tokens[2]));
```

Comment:
> merchant address is not P2PKH
I see you are validating this in another file, shouldn't you do the same to prevent creating such invalid contracts?